index.js

```diff
@@ -70,20 +72,71 @@ function format(text, opts, addAlignmentSize) {
   return str;
 }
 
-function formatRange(text, opts) {
+function findNodeByOffset(node, offset, opts, text) {
+  const start = util.locStart(node);
+  const end = util.locEnd(node);
+  if (start <= offset && offset <= end) {
+    for (const childNode of comments.getSortedChildNodes(node)) {
+      const childResult = findNodeByOffset(childNode, offset, opts, text);
+      if (childResult) {
+        return childResult;
+      }
+    }
+
+    try {
+      parser.parse(text.slice(start, end), opts);
+      return node;
+    } catch (err) {
+      return;
+    }
+  }
+}
+
+function calculateRange(text, opts, ast) {
+  const startNode = findNodeByOffset(ast, opts.rangeStart, opts, text);
+  const endNode = findNodeByOffset(ast, opts.rangeEnd, opts, text);
+  const rangeStart = Math.min(util.locStart(startNode), util.locStart(endNode));
+  const rangeEnd = Math.max(util.locEnd(startNode), util.locEnd(endNode));
+
+  const rangeString = text.slice(rangeStart, rangeEnd);
+
+  // Try to extend the range backwards to the beginning of the line.
+  // This is so we can detect indentation correctly and restore it.
+  const rangeAst = parser.parse(rangeString, opts);
   // Use `Math.min` since `lastIndexOf` returns 0 when `rangeStart` is 0
-  const rangeStart = Math.min(
-    opts.rangeStart,
-    text.lastIndexOf("\n", opts.rangeStart) + 1
+  const rangeStart2 = Math.min(
+    rangeStart,
+    text.lastIndexOf("\n", rangeStart) + 1
   );
-  // Use `text.length - 1` as the maximum since `indexOf` returns -1 if `fromIndex >= text.length`
-  const fromIndex = Math.min(opts.rangeEnd, text.length - 1);
-  const nextNewLineIndex = text.indexOf("\n", fromIndex);
-  const rangeEnd = (nextNewLineIndex < 0 ? fromIndex : nextNewLineIndex) + 1; // Add one to make rangeEnd exclusive
-
-  if (0 < rangeStart || rangeEnd < text.length) {
-    const rangeString = text.substring(rangeStart, rangeEnd);
-    const alignmentSize = getAlignmentSize(
+  const rangeString2 = text.slice(rangeStart2, rangeEnd);
+  try {
+    const rangeAst2 = parser.parse(rangeString2, opts);
+    if (cleanAST(rangeAst) === cleanAST(rangeAst2)) {
```

Comment:
The motivation was to restore the behavior that preserves indentation, even when the indentation isn't part of the range itself. Here's an example from https://github.com/prettier/prettier/pull/1659/commits/5d09c25166f4536586ddf014150d4f2753dc34db#diff-dc6e7b3e0a75f95cb931df4f9e698492R42:

Given this code:

```js
try {
  if (condition) {
    body
  }
}
catch (err) {}
```

and the range `[9, 37)` (this selects from the `f` in `if` to the `}` closing the if-body (inclusive)), then the command `prettier --range-start 8 --range-end 37` results in

```js
try {
  if (condition) {
    body;
  }
}
catch (err) {}
```

preserving the indent, whereas before this change, it would result in:

```js
try {
  if (condition) {
  body;
}
}
catch (err) {}
```

Maybe this isn't worth it, though, and we can just require that the range include the indentation if it is to be preserved. I'll revert this change (and its tests).